overlord/devicestate/handlers_systems.go

```diff
@@ -0,0 +1,307 @@
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package devicestate
+
+import (
+	"fmt"
+	"os"
+
+	"gopkg.in/tomb.v2"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/boot"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/overlord/assertstate"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/release"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/strutil"
+)
+
+func taskRecoverySystemSetup(t *state.Task) (*recoverySystemSetup, error) {
+	var setup recoverySystemSetup
+
+	err := t.Get("recovery-system-setup", &setup)
+	if err == nil {
+		return &setup, nil
+	}
+	if err != state.ErrNoState {
+		return nil, err
+	}
+	// find the task which holds the data
+	var id string
+	if err := t.Get("recovery-system-setup-task", &id); err != nil {
+		return nil, err
+	}
+	ts := t.State().Task(id)
+	if ts == nil {
+		return nil, fmt.Errorf("internal error: cannot find referenced task %v", id)
+	}
+	if err := ts.Get("recovery-system-setup", &setup); err != nil {
+		return nil, err
+	}
+	return &setup, nil
+}
+
+func setTaskRecoverySystemSetup(t *state.Task, setup *recoverySystemSetup) error {
+	if t.Has("recovery-system-setup") {
+		t.Set("recovery-system-setup", setup)
+		return nil
+	}
+	var id string
+	if err := t.Get("recovery-system-setup-task", &id); err != nil {
+		return err
+	}
+	ts := t.State().Task(id)
+	if ts == nil {
+		return fmt.Errorf("internal error: cannot find referenced task %v", id)
+	}
+	ts.Set("recovery-system-setup", setup)
+	return nil
+}
+
+func (m *DeviceManager) doCreateRecoverySystem(t *state.Task, _ *tomb.Tomb) error {
+	if release.OnClassic {
+		// TODO: this may need to be lifted in the future
+		return fmt.Errorf("cannot run update gadget assets task on a classic system")
+	}
+
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	remodelCtx, err := DeviceCtx(st, t, nil)
+	if err != nil {
+		return err
+	}
+	isRemodel := remodelCtx.ForRemodeling()
+	groundDeviceCtx := remodelCtx.GroundContext()
+
+	model := groundDeviceCtx.Model()
+	if isRemodel {
+		model = remodelCtx.Model()
+	}
+
+	setup, err := taskRecoverySystemSetup(t)
+	if err != nil {
+		return fmt.Errorf("internal error: cannot obtain recovery system setup information")
+	}
+	label := setup.Label
+
+	// get all infos
+	infoGetter := func(name string) (*snap.Info, bool, error) {
+		// snap may be present in the system in which case info comes
+		// from snapstate
+		info, err := snapstate.CurrentInfo(st, name)
+		if err == nil {
+			hash, _, err := asserts.SnapFileSHA3_384(info.MountFile())
```

Comment:
so we I understand we are going to hash the files twice? because of the Derive* on the other side of this? not an immediate blocker but we really need to think what else we can do

Reply:
There are some options for tweaking. For snaps being fetched we have download info in snapsup which carries the hash. For installed asserted snaps we could pull that from revision assertion. If I'm not mistaken, the last bit is unasserted snaps, which need to be hashed, as we store the hash in any way in the state.

Reply:
is this going to addressed in a follow-up? that's fine but want to keep track

Reply:
Yes, I don't want to derail this PR.